book-3-web-api/chapters/REACT_ROUTER.md

```diff
@@ -0,0 +1,334 @@
+                />
+              </FormGroup>
+              <FormGroup>
+                <Label for="title">Title</Label>
+                <Input id="title" onChange={(e) => setTitle(e.target.value)} />
+              </FormGroup>
+              <FormGroup>
+                <Label for="caption">Caption</Label>
+                <Input
+                  id="caption"
+                  onChange={(e) => setCaption(e.target.value)}
+                />
+              </FormGroup>
+            </Form>
+            <Button color="info" onClick={submit}>
+              SUBMIT
+            </Button>
+          </CardBody>
+        </Card>
+      </div>
+    </div>
+  );
+};
+
+export default PostForm;
+```
+
+Again, your Add form will likely look different than this, but there's only two sections to call out in this example. This code says to send the user back to the home (or `/`) route after the post has been successfully added.
+
+```js
+addPost(post).then((p) => {
+    // Navigate the user back to the home route
+    history.push("/");
+});
+```
+
+To get access to the `history` instance, we need to use the `useHistory` hook
+
+```js
+const history = useHistory();
+```
+
+Add these two lines of code to your own `PostForm` component and try adding a new post. You should be taken back to the feed after submitting the form.
+
+## Using URL params in Components
+
+The last thing we want to do with our new routing abilities, is create a `PostDetails` component and use the route parameter to decide which post's details we should be showing. For example, if a user navigates to `/posts/2`, the component code will have to read the route param of `2` and use that value to make a fetch call to get that post's details.
+
+Before we make a Post Details component, lets add a function to our provider that makes that fetch call
+
+> PostProvider.js
+
+```js
+const getPost = (id) => {
+    return fetch(`/api/post/${id}`).then((res) => res.json());
+};
+```
+
+Now make sure to provide that new function
+
+```js
+<PostContext.Provider value={{ posts, getAllPosts, addPost, getPost }}>
+```
+
+**NOTE** This assumes your API is set up to return a post object which includes an array of comments. If you need to make an additional fetch call to get the comments for a post, update the `getPost` function as needed.
+
+Now we can add a `PostDetails.js` file in your components directory. Notice the use of the `useParams` hook to access the route param.
+
+> PostDetails.js
+
+```js
+import React, { useEffect, useContext, useState } from "react";
+import { ListGroup, ListGroupItem } from "reactstrap";
+import { PostContext } from "../providers/PostProvider";
+import { useParams } from "react-router-dom";
+import Post from "./Post";
+
+const PostDetails = () => {
+  const [post, setPost] = useState();
+  const { getPost } = useContext(PostContext);
+  const { id } = useParams();
+
+  useEffect(() => {
+    getPost(id).then(setPost);
```

Comment:
This has me, once again, considering the way we use context to provide the entire list of posts vs how this code works. Is there a way to be consistent? does it make sense to have a "current post" in the `PostContext`?